// @ngx-gallery/core/esm5/public_api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export { Gallery } from './lib/services/gallery.service';
export { GalleryRef } from './lib/services/gallery-ref';
export { GalleryComponent } from './lib/components/gallery.component';
export { ImageItem, VideoItem, IframeItem, YoutubeItem } from './lib/components/templates/items.model';
export { GalleryIframeComponent } from './lib/components/templates/gallery-iframe.component';
export { GalleryImageComponent } from './lib/components/templates/gallery-image.component';
export { GalleryVideoComponent } from './lib/components/templates/gallery-video.component';
export { GALLERY_CONFIG } from './lib/models/config.model';
export {} from './lib/models/gallery.model';
export { GalleryAction, ImageSize, LoadingStrategy, ThumbnailsPosition, ImageLoaderMode, DotsPosition, CounterPosition, ThumbnailsMode, SlidingDirection, GalleryItemType } from './lib/models/constants';
export { GalleryModule } from './lib/gallery.module';
