// @ngx-gallery/core/esm2015/lib/components/templates/radial-progress.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
export class RadialProgressComponent {
    constructor() {
        this.mode = 'determinate';
        this.value = 0;
        this.diameter = 60;
        this.strokeWidth = 3;
    }
    /**
     * The radius of the spinner, adjusted for stroke width.
     * @return {?}
     */
    get circleRadius() {
        return (this.diameter - this.strokeWidth) / 2;
    }
    /**
     * The view box of the spinner's svg element.
     * @return {?}
     */
    get viewBox() {
        /** @type {?} */
        const viewBox = this.circleRadius * 2 + this.strokeWidth;
        return `0 0 ${viewBox} ${viewBox}`;
    }
    /**
     * The stroke circumference of the svg circle.
     * @return {?}
     */
    get strokeCircumference() {
        return 2 * Math.PI * this.circleRadius;
    }
    /**
     * The dash offset of the svg circle.
     * @return {?}
     */
    get strokeDashOffset() {
        if (this.mode === 'determinate') {
            return this.strokeCircumference * (100 - this.value) / 100;
        }
        // In fallback mode set the circle to 60% and rotate it with CSS.
        if (this.mode === 'indeterminate') {
            return this.strokeCircumference * 0.4;
        }
        return null;
    }
    /**
     * Stroke width of the circle in percent.
     * @return {?}
     */
    get circleStrokeWidth() {
        return this.strokeWidth / this.diameter * 100;
    }
}
RadialProgressComponent.decorators = [
    { type: Component, args: [{
                selector: 'radial-progress',
                changeDetection: ChangeDetectionStrategy.OnPush,
                host: {
                    '[attr.mode]': 'mode',
                    '[style.width.px]': 'diameter',
                    '[style.height.px]': 'diameter'
                },
                template: `
    <svg class="radial-progress"
         [style.width.px]="diameter"
         [style.height.px]="diameter"
         [attr.viewBox]="viewBox"
         preserveAspectRatio="xMidYMid meet"
         focusable="false">

      <circle class="radial-progress-meter"
              cx="50%"
              cy="50%"
              [attr.r]="circleRadius"
              [style.stroke-width.%]="circleStrokeWidth"/>

      <circle class="radial-progress-value"
              cx="50%"
              cy="50%"
              [attr.r]="circleRadius"
              [style.stroke-dashoffset.px]="strokeDashOffset"
              [style.stroke-dasharray.px]="strokeCircumference"
              [style.stroke-width.%]="circleStrokeWidth"/>
    </svg>
  `
            }] }
];
RadialProgressComponent.propDecorators = {
    mode: [{ type: Input }],
    value: [{ type: Input }],
    diameter: [{ type: Input }],
    strokeWidth: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    RadialProgressComponent.prototype.mode;
    /** @type {?} */
    RadialProgressComponent.prototype.value;
    /** @type {?} */
    RadialProgressComponent.prototype.diameter;
    /** @type {?} */
    RadialProgressComponent.prototype.strokeWidth;
}
